### PR TITLE
W/ 2.0 i2c integration

### PR DESCRIPTION
**Pull request by kfirmanty**
#### What does this PR do?
Hi,

This PR will introduce operators which will handle i2c communication with 3 new modes in W/ 2.0 module - delay, synth and tape. Some of the things still need some work - like retrieving values from W/ - but I created this PR so others could comment on it already or even maybe we could work on this integration together if someone is willing and has free time to do so.

#### Provide links to any related discussion on [lines](https://llllllll.co/).
[lines topic](https://llllllll.co/t/teletype-and-w-2-0-integration/34220)

#### How should this be manually tested?
Most of these operators could be tested by calling them with minimum possible value and maximum and checking if lights on W/ module are updated accordingly. 

#### Needs to be done
* [ ]  better naming
* [ ]  retrieving values from i2c - currently commands are only sent from teletype to W/
#### I have,
* [ ] updated `CHANGELOG.md`
* [ ] updated the documentation
* [ ] run `make format` on each commit

### CODE REVIEW

File: src/ops/i2c.h [hunk 1]
```diff
@@ -26,4 +26,35 @@ extern const tele_op_t op_IIBB1;
 extern const tele_op_t op_IIBB2;
 extern const tele_op_t op_IIBB3;
 
+extern void i2c_get_0(command_state_t *cs, uint8_t addr, uint8_t cmd);
```
Comment (scanner-darkly) [inline src/ops/i2c.h:29]: are these generic i2c helpers? i would put them into a separate file under src - this one is for the generic i2c ops.

**Comment by scanner-darkly** [comment]
other than the 1 comment looks great!

**Comment by trentgill** [comment]
The compilation doesn't seem to be working. Is it working on your system? Seems like you might have made changes to libavr32 but haven't pushed the changes to that submodule (WS_ADDR_6 is undefined).

**Comment by kfirmanty** [comment]
Yeah, sorry about that I have defined few constants in `ii.h` of `libavr32`:
```
--- a/src/ii.h
+++ b/src/ii.h
@@ -66,6 +66,12 @@
 #define WS_ADDR_2           0x72
 #define WS_ADDR_3           0x73
 #define WS_ADDR_4           0x74
+#define WS_ADDR_5           0x75
+#define WS_ADDR_6           0x76
+#define WS_ADDR_7           0x77
+#define WS_ADDR_8           0x78
+#define WS_ADDR_9           0x79
+

 // COMMANDS
 // new commands can overlap with existing ones
@@ -212,3 +218,52 @@
 #define WS_PLAY 2
 #define WS_LOOP 3
 #define WS_CUE  4
+
+#define WS_S_VEL   2
+#define WS_S_PITCH   3
+#define WS_S_VOX  4
+#define WS_S_NOTE  5
+#define WS_S_AR_MODE  7
+#define WS_S_CURVE   8
+#define WS_S_RAMP   9
+#define WS_S_FM_INDEX  10
+#define WS_S_FM_RATIO  11
+#define WS_S_LPG_TIME  12
+#define WS_S_LPG_SYMMETRY  13
+#define WS_S_PATCH  14
+#define WS_S_VOICES  15
+#define WS_S_FM_ENV  16
+
+#define WS_D_FEEDBACK  2
+#define WS_D_MIX   3
+#define WS_D_LOWPASS  4
+#define WS_D_FREEZE   5
+#define WS_D_TIME    6
+#define WS_D_LENGTH    7
+#define WS_D_POSITION    8
+#define WS_D_CUT    9
+#define WS_D_FREQ_RANGE   10
+#define WS_D_RATE    11
+#define WS_D_FREQ    12
+#define WS_D_CLK  13
+#define WS_D_CLK_RATIO   14
+#define WS_D_PLUCK   15
+#define WS_D_MOD_RATE   16
+#define WS_D_MOD_AMOUNT   17
+
+#define WS_T_RECORD   1
+#define WS_T_PLAY   2
+#define WS_T_REV 3
+#define WS_T_SPEED   4
+#define WS_T_FREQ   5
+#define WS_T_PRE_LEVEL  6
+#define WS_T_MONITOR_LEVEL  7
+#define WS_T_REC_LEVEL  8
+#define WS_T_HEAD_ORDER  9
+#define WS_T_LOOP_START 10
+#define WS_T_LOOP_END  11
+#define WS_T_LOOP_ACTIVE  12
+#define WS_T_LOOP_SCALE  13
+#define WS_T_LOOP_NEXT  14
+#define WS_T_TIMESTAMP   15
+#define WS_T_SEEK  16
```

**Comment by tehn** [comment]
you'll have to first submit a PR for your changes to libavr32

**Comment by kfirmanty** [comment]
Hi,
I did a fresh checkout for this branch and build it on local machine just fine using:
```
docker run --rm -it -v"$(pwd)":/target dewb/monome-build bash
make
```
does travis build needs any additional configuration?

**Comment by scanner-darkly** [comment]
you also need to link  your op files [here](https://github.com/monome/teletype/blob/main/simulator/Makefile) and [here](https://github.com/monome/teletype/blob/main/tests/Makefile)

**Comment by kfirmanty** [comment]
> you also need to link your op files [here](https://github.com/monome/teletype/blob/main/simulator/Makefile) and [here](https://github.com/monome/teletype/blob/main/tests/Makefile)

Thanks! I updated these files so hopefully Travis will be happy now ;) Of course I will update documentation and will write some tests probably later. But if anyone is willing to do rest of the work needed for this integration to work in 100% then please free to do so in that branch, I don't want to block any progress.

**Comment by tehn** [comment]
@trentgill considering a TT release soon, thought you'd like this in there

**Comment by scanner-darkly** [comment]
@kfirmanty are you able to complete this pull request? the outstanding items:
- rebase / fix conflicts
- documentation
- teletype help

**Comment by trentgill** [comment]
I’d love this to be included but i wont have the time to jump on any code bits for a few weeks. If it makes it in that’d be great, otherwise i’ll polish it up ahead of the next release.

**Comment by scanner-darkly** [comment]
i can help with rebasing / fixing conflicts, and if somebody can provide the actual text i can add help/docs.

**Comment by kfirmanty** [comment]
> @kfirmanty are you able to complete this pull request? the outstanding items:
> 
> * rebase / fix conflicts
> * documentation
> * teletype help

Hi @scanner-darkly right now my job is taking all of my time, but I will try maybe over the weekend to fix outstanding issues. Sorry that it takes so long.

**Comment by desolationjones** [comment]
Based on @trentgill's [comment](https://github.com/monome/libavr32/pull/57#issuecomment-664680099), I think `libavr` needs the I2C address assignments updated. 0x73 and 0x74 should be reserved, 0x75 should be `JF_ADDR_2`. `WS_ADDR_*` will only go up to 6. Should I create a separate pull request for `libavr` before these TT changes are rebased?

**Comment by desolationjones** [comment]
I opened a [PR on libavr](https://github.com/monome/libavr32/pull/61) with the updated I2C address definitions. I changed the nomenclature a tad bit to make the address names match the W/2.0 modes. I think whoever is taking this W/2.0 TT change across the finish line should bundle my `ii.h` change from libavr and we can just close that PR.

**Comment by tehn** [comment]
@trentgill this looks good to go now, can you double-check?

**Review by trentgill — APPROVED** [review]
This all looks good to me. As long as someone has tested it, I think it's good to go.

**Pull request merged**